rust-runtime/aws-smithy-wasm/src/wasi.rs

```diff
@@ -0,0 +1,314 @@
+                    SdkBody::empty()
+                } else {
+                    SdkBody::from(body)
+                }
+            });
+            tracing::trace!("WasiHttpConnector: response received {response:?}");
+
+            let sdk_res = Response::try_from(response)
+                .map_err(|err| ConnectorError::other(err.into(), None))?;
+
+            Ok(sdk_res)
+        })
+    }
+}
+
+/// Wasi http client containing the options to pass to the
+struct WasiDefaultClient {
+    options: WasiRequestOptions,
+}
+
+impl WasiDefaultClient {
+    /// Create a new Wasi HTTP client.
+    fn new(options: WasiRequestOptions) -> Self {
+        Self { options }
+    }
+
+    /// Make outgoing http request in a Wasi environment
+    fn handle(&self, req: http::Request<Bytes>) -> Result<http::Response<Bytes>, ConnectorError> {
+        let req = WasiRequest::try_from(req).expect("Converting http request");
+
+        let res = outgoing_handler::handle(req.0, self.options.clone().0).expect("Http response");
+
+        // Right now only synchronous calls can be made through WASI, so we subscribe and
+        // block on the FutureIncomingResponse
+        let subscription = res.subscribe();
+        subscription.block();
+
+        //The FutureIncomingResponse .get() method returns a
+        //Option<Result<Result<IncomingResponse, ErrorCode>, ()>>.
+        //The outer Option ensures readiness which we know is Some because we .block() waiting for it
+        //The outer Result is just a singleton enforcer so we can only get the response once
+        //The inner Result indicates whether the HTTP call was sent/received successfully (not the 200 succes of the call)
+        let incoming_res = res
+            .get()
+            .expect("Http response not ready")
+            .expect("Http response accessed more than once")
+            .map_err(|err| ConnectorError::other(err.into(), None))?;
+
+        let response = http::Response::try_from(WasiResponse(incoming_res))
+            .expect("Converting to http response");
+
+        Ok(response)
+    }
+}
+
+/// Wrapper for the Wasi RequestOptions type to allow us to impl Clone
+struct WasiRequestOptions(Option<outgoing_handler::RequestOptions>);
+
+//The Wasi RequestOptions type doesn't impl copy or clone but the outgoing_handler::handle method
+//takes ownership, so we impl it on this wrapper type
+impl Clone for WasiRequestOptions {
+    fn clone(&self) -> Self {
+        let new_opts = if let Some(opts) = &self.0 {
+            let new_opts = RequestOptions::new();
+            new_opts
+                .set_between_bytes_timeout(opts.between_bytes_timeout())
+                .expect("Between bytes timeout");
+            new_opts
+                .set_connect_timeout(opts.connect_timeout())
+                .expect("Connect timeout");
+            new_opts
+                .set_first_byte_timeout(opts.first_byte_timeout())
+                .expect("First byte timeout");
```

Comment:
So are all the expects due to each call out to a binding potentially failing due to a function not existing or something?

Reply:
No, each of these `set_some_http_attribute` methods can fail if the string passed in can't be correctly parsed as the expected attribute. You can see the docs for those methods [here](https://docs.rs/wasi/0.12.0+wasi-0.2.0/wasi/http/types/struct.OutgoingRequest.html#method.set_path_with_query). Thankfully since this is a `Clone` we know that the values from the existing value should always be valid in this context.